staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -0,0 +1,211 @@
+	fcmetrics "k8s.io/apiserver/pkg/util/flowcontrol/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const (
+	decisionExempt = iota
+	decisionExecute
+	decisionCancelWait
+	decisionReject
+	decisionError
+)
+
+type fakeApfFilter struct {
+	mockDecision int
+}
+
+func (t fakeApfFilter) Handle(ctx context.Context,
+	requestDigest utilflowcontrol.RequestDigest,
+	noteFn func(fs *fctypesv1a1.FlowSchema, pl *fctypesv1a1.PriorityLevelConfiguration),
+	queueNoteFn fq.QueueNoteFn,
+	execFn func(),
+) {
+	if t.mockDecision == decisionError {
+		panic("Handle should not be invoked")
+	}
+	noteFn(bootstrap.SuggestedFlowSchemaGlobalDefault, bootstrap.SuggestedPriorityLevelConfigurationGlobalDefault)
+	switch t.mockDecision {
+	case decisionExempt:
+		execFn()
+	case decisionExecute:
+		queueNoteFn(true)
+		queueNoteFn(false)
+		execFn()
+	case decisionCancelWait:
+		queueNoteFn(true)
+		queueNoteFn(false)
+	case decisionReject:
+		return
+	}
+}
+
+func (t fakeApfFilter) Run(stopCh <-chan struct{}) error {
+	return nil
+}
+
+func (t fakeApfFilter) Install(c *mux.PathRecorderMux) {
+}
+
+func newApfServer(decision int) *httptest.Server {
+	requestInfoFactory := &apirequest.RequestInfoFactory{APIPrefixes: sets.NewString("apis", "api"), GrouplessAPIPrefixes: sets.NewString("api")}
+	longRunningRequestCheck := BasicLongRunningRequestCheck(sets.NewString("watch"), sets.NewString("proxy"))
+
+	apfHandler := WithPriorityAndFairness(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	}), longRunningRequestCheck, fakeApfFilter{mockDecision: decision})
+	handler := apifilters.WithRequestInfo(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		r = r.WithContext(apirequest.WithUser(r.Context(), &user.DefaultInfo{
+			Groups: []string{user.AllUnauthenticated},
+		}))
+		apfHandler.ServeHTTP(w, r)
+	}), requestInfoFactory)
+
+	apfServer := httptest.NewServer(handler)
+	return apfServer
+}
+
+func TestApfSkipLongRunningRequest(t *testing.T) {
+	epmetrics.Register()
+
+	server := newApfServer(decisionError)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces?watch=true", server.URL), http.StatusOK); err != nil {
+		// request should not be rejected
+		t.Error(err)
+	}
+
+	checkForExpectedMetrics(t, []string{
+		"apiserver_current_inflight_requests",
+	})
+}
+
+func TestApfRejectRequest(t *testing.T) {
+	epmetrics.Register()
+
+	server := newApfServer(decisionReject)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusTooManyRequests); err != nil {
+		t.Error(err)
+	}
+
+	// wait metrics to be collected
+	time.Sleep(time.Millisecond * 1500)
+
+	checkForExpectedMetrics(t, []string{
+		"apiserver_request_terminations_total",
+		"apiserver_dropped_requests_total",
+	})
+}
+
+func TestApfExemptRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+
+	// wait ReadWriteConcurrencyObserverPairGenerator to be registered
```

Comment:
The wait here is not for an asynchronous process.  It is for time to pass so that the loop at https://github.com/kubernetes/apiserver/blob/09eab1cf2cce55a46e5d9042c90f136d91383e19/pkg/util/flowcontrol/metrics/sample_and_watermark.go#L185 will go through a non-zero number of iterations.